### ISSUE TITLE
Support for ESLint v8

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#3738]
ESLint has just released the first beta version of v8.

https://eslint.org/blog/2021/08/eslint-v8.0.0-beta.0-released

This version contains a number of breaking changes which will prevent our tooling from working at all with it. This is expected and we had discussed the changes with the ESLint team as they worked on v8.

This project is maintained by volunteers. Support will be added as time permits.  
If you're not willing or able to help with adding support:
- Please do not raise issues.
- Please do not comment on threads.
- Please be patient and subscribe to this issue for updates.

This update will require breaking changes on our side as well, so it will take some time to support it.
There is no ETA on this work. Again *please be patient*.

----

# Breaking Changes in ESLint v8

## Breaking: Strict package exports

**PRs:** #3736, #3737

ESLint has restricted what can be imported from their package. In doing so they have continued to export some of the things we need to operate, though they are available on a different path.

*This will require a large change in how the package is architected.*

**Problem:** How can we support both <v8 and >=v8 at the same time?

## Breaking: Require meta.hasSuggestions for rules with suggestions

**PRs:** #3737

This will require updates to some of our rule defs as we did not annotate this after the release of the flag some time in ESLint v7.

## Breaking: fix comma-dangle schema

**PRs:** None yet.

We will have to update our extension rule's schema.

## Breaking: upgrade espree and support new class features

**PRs:** #2933
**Related Issues:**  #3430

We supported the old babel-style AST for class properties (`ClassProperty`). ESLint has added support for properties using ESTree's official spec. We need to update our representation as appropriate.

## Breaking: add some rules to eslint:recommended

**PRs:** None yet.

We will want to review the changes and update our lists as appropriate.
For example - ESLint now recommends [`no-unsafe-optional-chaining`](https://eslint.org/docs/rules/no-unsafe-optional-chaining). Every single check of this rule is [covered by typescript](https://www.typescriptlang.org/play?target=99&ts=4.0.5#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV1ARgAoIAjAKwC44UkAbWuAHzgG8EIIrCBKOAXgB8cAG4RMAEwC+vFgFgAUHDjFyAfgB07CNx4BuBZIULQkWIlQZsuUACYVlanQbM2HKizgkAhlCoByPzhJaVYFJXsNLW51byh9eUN5Y3BoeGR0LBw8EABmQi0qGnomZR5+IRZYqgBnGChMFABzaRkw5S0NHmjY+MTk0zSLTOsQABZ8t0di5kIywVYq5V550QkW0MV2jk7ubp89AyN5E1TzDKtsgFZ7QqcS104p51LloVXxbgAaTx8uV5ExFJWpsIpoOHBVKo4KQyBpYjpuL0jiczOlLFlQAA2CaPIrPWb-d5faHkW7TVgFJblAFrYHhLRMZgwyIcBFIpLHFKooYXUAAdhuT3ulJIHFowC8KEk30pczegLpymZYIgcAAZGrEKyDglkVzBucMSAABw4qiiiDiyXfGFk54PP7U97Sn6+KkrBUbengqHKhnmnxsw4clEG9EjACcZsQXlo1WANtJQpci0qXgAXv4-MFFTjE7CYoHC+n2f1Tmjhtl8AAGQV44WTQjqZs+RrVKiSgCe-1q9SaITkIL9rIABjBgABbMC0LzjkelzkDM7hqtENAz9OZ5MUyYsFDAADuPCoAGVgDAADy9hqNAQ5r3UQ-KdcZ9Ms7Q6vqL8s8o34Oy2tuDpwNeTQANoALoDm0YHNuow4QBBC6hsulYEHkgH1i4lKgY0kEuosTYtlAbY1HUN6QYSnqDkosREfBagqoiwZltyhojPg4yYXc2GTLh+GKvgcANCSBZaMh+qobyID4Nc3HksBMCdmAwAQAgcBnjABG-HAqAANYoBAB4oIqsTCSgtSSmgqnqQhElLhW0n4Ni8n2jh5HgVBirivAPRtOwUDKGZamie+zG6iGkmOX+AquQ2jweIsARBNBmxoDgtSsK6QT8KFKrxEoUDnkgUC4H5EWsWGaEyaacW8QlSXZt556ugVcBgYlPg5XweVaBBuWQW1RUwCVZU+PZP7sVWUZ1TujxKSpIWaalSgvtU1RwAAgtk44oOIG0IawiRfih0UjDYtazYpyk2Rp54rSIXVeLla0bdtoC7ftvXgiwkhDcVpVwF4E1sSuthEFdlIAApQBAE6YPGF4sMeNLiN8jWSHeipeNUnboFJWQMmUNFKHA6UWfAnWBZIuVeAeXiYPAdltOEdMM0zjFRDqpPKGzjPfdohZxG0iSFQDuDiSx8hAA)

## Breaking: drop node v10/v13/v15

**PRs:** None yet.

We will want to update our node ranges as appropraite.

## Changes with no impact:

- Breaking: disallow SourceCode#getComments() in RuleTester
  - This is not a problem. We could probably remove the API from our types now.
- Breaking: Fixable disable directives
- Breaking: require meta for fixable rules
- Breaking: allow all directives in line comments
- Breaking: support new regex d flag
- Breaking: Drop codeframe and table formatters

**Comment by bradzacher** [comment]
updated for the beta.2 release - no changes need addressing.

**Comment by SimenB** [comment]
I don't see it mentioned here - which versions of ESLint will v5 support? 6, 7 and 8 or something else? 🙂

**Comment by MichaelDeBoey** [comment]
@SimenB It's mentioned in #3774

> Drop Support for ESLint v5
> * ESLint v5 is now 3 years old - why are you still using it? Upgrade to at least ESLint v6.

**Comment by SimenB** [comment]
Ah, thanks! Missed that PR

(I'll hide my comments)

**Comment by bradzacher** [comment]
@SimenB yup we'll be dropping support for v5 in our v5 because it's so old now and it hampers our support of new extension rules.

TBH I'd love to drop support for v6... but I know people are "stuck" on that version due to the number of breaking changes v7 introduced.

Well, not stuck, rather they don't have bandwidth to fix all of the issues caused by the breaking changes.
For example I know first-hand FB is stuck on v6 because it's a big task to fix the 500+ internal lint rules to be compatible with v7 (I've looked... It's going to be *days* if not ***weeks*** of effort 😭).

**Comment by Brian-Bartels** [comment]
Anyone tried the ESLint complexity rule yet? It has worked in the past. But I think that they made some changes in v8 that might be incompatible. I wasn't getting a `codepath` object in the `oncodepathend` function. Note that a copy paste of the previous version of ESLint complexity still runs normally even on v5 parser and v8 ESLint

**Comment by bradzacher** [comment]
@Brian-Bartels If you can create a standalone repro which shows the rule working with vanilla ESLint v8 but not working with ESLint v8 plus our parser v5 - please file an issue so we can look into it.

**Comment by Brian-Bartels** [comment]
@bradzacher Well I made a repro attempt. And it still works. So there must of been something wrong with my configurations before. Proof that it works here https://github.com/Brian-Bartels/complexityRepro

**Comment by nicolas377** [comment]
You all might want to release this soon. Eslint v8 just hit npm.

**Comment by vtrphan** [comment]
i'm having this error too now that i've upgraded to esline v8 and am waiting for this new release of this plugin

**Comment by nicolas377** [comment]
v4.33 doesn't have support for eslint v8. Try installing from the rc-v5 tag, and try again!
